distributed/client.py

```diff
@@ -3447,32 +3446,32 @@ def compute(
             )
 
         variables = [a for a in collections if dask.is_dask_collection(a)]
+        if newstyle_collections(variables):
+            variables = [var.finalize_compute() for var in variables]
+            dsk = collections_to_dsk(variables, optimize_graph, **kwargs)
+        else:
+            dsk = collections_to_dsk(variables, optimize_graph, **kwargs)
+            names = ["finalize-%s" % tokenize(v) for v in variables]
+            dsk = dsk._hlg
+            dsk2 = {}
+            for i, (name, v) in enumerate(zip(names, variables)):
+                func, extra_args = v.__dask_postcompute__()
+                keys = v.__dask_keys__()
+                if func is single_key and len(keys) == 1 and not extra_args:
+                    names[i] = keys[0]
+                else:
+                    dsk2[name] = (func, keys) + extra_args
 
-        dsk = self.collections_to_dsk(variables, optimize_graph, **kwargs)
-        names = ["finalize-%s" % tokenize(v) for v in variables]
-        dsk2 = {}
-        for i, (name, v) in enumerate(zip(names, variables)):
-            func, extra_args = v.__dask_postcompute__()
-            keys = v.__dask_keys__()
-            if func is single_key and len(keys) == 1 and not extra_args:
-                names[i] = keys[0]
-            else:
-                dsk2[name] = (func, keys) + extra_args
-
-        if not isinstance(dsk, HighLevelGraph):
-            dsk = HighLevelGraph.from_collections(id(dsk), dsk, dependencies=())
-
-        # Let's append the finalize graph to dsk
-        finalize_name = tokenize(names)
-        layers = {finalize_name: dsk2}
-        layers.update(dsk.layers)
-        dependencies = {finalize_name: set(dsk.layers.keys())}
-        dependencies.update(dsk.dependencies)
-        dsk = HighLevelGraph(layers, dependencies)
+            # Let's append the finalize graph to dsk
+            finalize_name = tokenize(names)
+            layers = {finalize_name: dsk2}
+            layers.update(dsk.layers)
+            dependencies = {finalize_name: set(dsk.layers.keys())}
+            dependencies.update(dsk.dependencies)
+            dsk = TaskFactoryHLGWrapper(HighLevelGraph(layers, dependencies), out_keys=names)
```

Comment:
The synchronous version is a little simpler and can be seen  here https://github.com/dask/dask/pull/10369/files#r1432858069